Packages/com.unity.inputsystem/InputSystem/Plugins/iOS/iOSStepCounter.cs

```diff
@@ -0,0 +1,166 @@
+#if UNITY_EDITOR || UNITY_IOS || UNITY_TVOS
+using System.Runtime.InteropServices;
+using AOT;
+using Unity.Collections.LowLevel.Unsafe;
+using UnityEngine.InputSystem.Layouts;
+using UnityEngine.InputSystem.LowLevel;
+using UnityEngine.InputSystem.Utilities;
+
+namespace UnityEngine.InputSystem.iOS.LowLevel
+{
+    //See CoreMotion.framework/Headers/CMAuthorization.h
+    public enum MotionAuthorizationStatus : int
+    {
+        NotDetermined = 0,
+        Restricted,
+        Denied,
+        Authorized
+    }
+
+    [StructLayout(LayoutKind.Sequential)]
+    internal unsafe struct iOSStepCounterState : IInputStateTypeInfo
+    {
+        public static FourCC kFormat = new FourCC('I', 'S', 'C', 'S');
+        public FourCC format => kFormat;
+
+        [InputControl(name = "stepCounter")]
+        public int stepCounter;
+    }
+
+    /// <summary>
+    /// Step Counter (also known as pedometer) sensor for iOS.
+    /// Note: You need to add 'Privacy - Motion Usage Description' to Info.plist to use this sensor.
+    /// </summary>
+    [InputControlLayout(stateType = typeof(iOSStepCounterState), variants = "StepCounter", hideInUI = true)]
+    public class iOSStepCounter : StepCounter
+    {
+        private const int kCommandFailure = -1;
+        private const int kCommandSuccess = 1;
+
+        internal delegate void OnDataReceivedDelegate(int deviceId, int numberOfSteps);
+
+        [StructLayout(LayoutKind.Sequential)]
+        private struct iOSStepCounterCallbacks
+        {
+            internal OnDataReceivedDelegate onData;
+        }
+
+        [DllImport("__Internal")]
+        private static extern int _iOSStepCounterEnable(int deviceId, ref iOSStepCounterCallbacks callbacks, int sizeOfCallbacks);
+
+        [DllImport("__Internal")]
+        private static extern int _iOSStepCounterDisable(int deviceId);
+
+        [DllImport("__Internal")]
+        private static extern int _iOSStepCounterIsEnabled(int deviceId);
+
+        [DllImport("__Internal")]
+        private static extern int _iOSStepCounterIsAvailable();
+
+        [DllImport("__Internal")]
+        private static extern int _iOSStepCounterGetAuthorizationStatus();
+
+        [MonoPInvokeCallback(typeof(OnDataReceivedDelegate))]
+        private static void OnDataReceived(int deviceId, int numberOfSteps)
+        {
+            var stepCounter = (iOSStepCounter)InputSystem.GetDeviceById(deviceId);
```

Comment:
I guess input system native API is not reachable from Obj-C? Otherwise it would probably be easier to just go directly to it instead of dealing with native->managed->native double transition.

Reply:
Yes, it's not reachable, but I don't understand where you see the last entry to native, you mean  InputSystem.GetDeviceById goes into native?

Reply:
- Your Obj-C code invokes manage code
- Your managed code queues event, which invokes native code

Reply:
Ah, I see... Thanks for explanation